### ISSUE TITLE
jack-in generates invalid shadow-cljs call

### ISSUE THREAD

**Issue by zilti** [BetterThanTomorrow/calva#2239]
Calva generates an unsupported shadow-cljs call for project type `shadow-cljs`, see [this issue report](https://github.com/thheller/shadow-cljs/issues/1129) on the shadow-cljs repository. For some reason, the generated command line still works on macOS and Linux, but it does not work on Windows.

**Comment by PEZ** [comment]
Hi! Can you share the steps you take to get that command line generated? I don't think know how to make it add `clj` aliases...

**Comment by zilti** [comment]
I can, yes, here it is:

```json
"calva.replConnectSequences": [
    {
      "name": "redsky Server",
      "projectType": "deps.edn",
      "cljsType": "none",
      "menuSelections": {
        "cljAliases": [
          "dev/deps",
          "dev/env"
        ]
      }
    },
    {
      "name": "redsky Server + Client",
      "projectType": "shadow-cljs",
      "cljsType": "shadow-cljs",
      "menuSelections": {
        "cljAliases": [
          "dev/deps",
          "dev/env"
        ],
        "cljsLaunchBuilds": [
          "app",
          "portfolio"
        ],
        "cljsDefaultBuild": "app"
      }
    }
  ]
```

This works on Linux and MacOS, but not on Windows, for some reason.

Alternatively, I also tried to use `"projectType": "deps.edn"` instead, but for some reason, that does not work - even though, when choosing "deps.edn + shadow-cljs" in the jack-in menu and selecting the exact same aliases and builds, it works fine. So something fails there when "scripting" it. Or I'd have to do something different to achieve the same.

**Comment by PEZ** [comment]
Thanks! Which version of Calva are you using?

**Comment by zilti** [comment]
According to the output when running it, it's v2.0.374.

**Comment by PEZ** [comment]
Hi again, sorry for the radio silence. I've hadn't had the time yet to try reproduce this. I'm pretty sure I can and that we will be able to fix the issue pretty easily.

Meanwhile: @thheller mentioned, in [that other issue](https://github.com/thheller/shadow-cljs/issues/1129), that you can provide the clj aliases in the shadow-cljs file. You would change from `:deps true` in `shadow-cljs.edn` to `:deps ["dev/deps", "dev/env"]`, and remove the `"cljAlisases"` configuration from that connect sequence. This should work around the issue.

**Comment by zilti** [comment]
Thank you a lot! Unfortunately, that workaround does not work for us, since we already have an alias in there, and adding `dev/env` would break our build due to it adding JVM startup options for dev only.

Anyway, we did find a different workaround for now, also mentioned by @thheller which is setting the customJackInCommandLine to 

`clojure -Sdeps '{:deps {cider/cider-nrepl {:mvn/version \"0.28.5\"}}}' -A:dev/deps:dev/env -M -m shadow.cljs.devtools.cli watch app portfolio`.

**Comment by PEZ** [comment]
Thanks for updating about the workaround not working. Helps me prioritize the issue.

Great that you found a workaround. It should work with a custom command line like so too:

```
npx.cmd shadow-cljs -d cider/cider-nrepl:0.28.5 -A :dev/deps:dev/env watch app portfolio
```

It's a matter of preference where you like the shadow-cljs output to go.

In that other issue you mentioned that `deps.edn + shadow-cljs` didn't work for you. Can you file an issue about it? I'd like to try reproduce it and fix that too.

**Comment by PEZ** [comment]
This should be fixed in this dev build:

* https://output.circle-artifacts.com/output/job/4ad2128a-2c29-4418-b4c8-a9f58355ef8e/artifacts/0/tmp/artifacts/calva-2.0.378-2242-deps-shadow-custom-sequence-f896c78d.vsix

Can you test it for us on Windows, @zilti?

**Comment by thheller** [comment]
I don't think this is going to work, and likely broke all other hosts too?

`args` seems to be supposed to be an array and each element corresponding to one argument. But now the `"-A :foo"` result is one string. If the node execution works like any other (e.g. JVM) that will end up getting escaped and arrive as one argument on the shadow-cljs end. I didn't actually test, but based on the fact that there is as `args` array getting constructed I presume the desired correct result should be `["-A", ":foo"]` in the array, not `["-A :foo"]`.

**Comment by PEZ** [comment]
Thanks for having a look, @thheller! 🙏 It generates this command line on Mac:

```
pushd /Users/pez/Projects/calva/test-data/projects/shadow-w-backend ; npx shadow-cljs -d cider/cider-nrepl:0.28.5 -A :dev:dev-too watch app app-too ; popd
```

Which works. But you could very well be right that it doesn't help with the issue on Windows.

**Comment by thheller** [comment]
The problem is that "printed" this command line appears totally normal, but at least on the JVM an arglist of `["-A" ":dev:dev-too"]` will have a different meaning than `["-A :dev:dev-too"]`. That might very well only be on issue on Windows, I'm not sure but I'd recommend following the API properly and properly diving arguments in the array.

**Comment by PEZ** [comment]
Indeed. I now fired up my Windows test machine, and it was still failing there. Adding the flag and the aliases to the args array separately works. Thanks again for the holler!

**Comment by PEZ** [comment]
I updated the docs around this a bit, as it was a bit unclear. They now look like so:

https://deploy-preview-2254--calva-docs.netlify.app/connect/#shadow-cljs-in-full-stack-projects

I hope they make more sense now...

**Comment by PEZ** [comment]
Fixed in 2.0.239.

**Comment by zilti** [comment]
Thanks!
